### PR TITLE
Check passcode to open app

### PR DESCRIPTION
**Pull request by biafra23**
Handles #1354 

Changes proposed in this pull request:
- Show fingerprint icon if biometrics enabled in settings
- Check passcode to open app
- Check passcode when opening from notification
  - Do not check passcode if app was already open whn notification was used to open the app
- Check passcode when resuming app from background
- "Forgot passcode" deletes all safes and secret keys
- **Missing**: Biometrics part (separate PR)

@gnosis/mobile-devs

### CODE REVIEW

File: app/src/main/java/io/gnosis/safe/HeimdallApplication.kt [hunk 1]
```diff
@@ -24,9 +26,60 @@ class HeimdallApplication : MultiDexApplication(), ComponentProvider {
...
     override fun get(): ApplicationComponent = component
 
+    val activityListeners = mutableListOf<OnActivityChanged>()
+    fun registerForActivity(listener: OnActivityChanged) {
+        activityListeners.add(listener)
+    }
+
     override fun onCreate() {
         super.onCreate()
+        registerActivityLifecycleCallbacks(object : ActivityLifecycleCallbacks {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/HeimdallApplication.kt:36]: can this be moved to a private method to make onCreate more legible?

File: app/src/main/java/io/gnosis/safe/HeimdallApplication.kt [hunk 2]
```diff
@@ -50,6 +103,11 @@ class HeimdallApplication : MultiDexApplication(), ComponentProvider {
     }
 }
 
+interface OnActivityChanged {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/HeimdallApplication.kt:106]: I would look for a better naming here. OnActivityChanged -> AppStateListener?

File: app/src/main/java/io/gnosis/safe/helpers/AddressInputHelper.kt [hunk 3]
```diff
@@ -50,6 +51,7 @@ class AddressInputHelper(
                 }
                 bottomSheetAddressInputQrTouch.setOnClickListener {
                     QRCodeScanActivity.startForResult(fragment)
+                    (fragment.activity as? StartActivity)?.comingFromBackground = true
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/helpers/AddressInputHelper.kt:54]: why do you have to set it manually?

File: app/src/main/java/io/gnosis/safe/ui/StartActivity.kt [hunk 4]
```diff
@@ -58,6 +69,8 @@ class StartActivity : BaseActivity(), SafeOverviewNavigationHandler {
         setupNav()
 
         handleNotifications(intent)
+
+        (application as? HeimdallApplication)?.registerForActivity(this)
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/StartActivity.kt:73]: should we move it to a base activity? (OnActivityChanged could be also implemented by a base activity)

File: app/src/main/java/io/gnosis/safe/ui/StartActivity.kt [hunk 5]
```diff
@@ -32,24 +34,33 @@ import pm.gnosis.utils.asEthereumAddress
...
     private val toolbar by lazy { findViewById<View>(R.id.toolbar) }
     private val navBar by lazy { findViewById<BottomNavigationView>(R.id.nav_bar) }
 
+    var comingFromBackground = false
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_start)
 
+        if (settingsHandler.requireToOpen && settingsHandler.usePasscode) {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/StartActivity.kt:57]: requireToOpen is ambiguous. Rename to requireCredentialsToOpen?

File: app/src/main/java/io/gnosis/safe/ui/StartActivity.kt (hunk 4 above)
Comment (biafra23) [reply app/src/main/java/io/gnosis/safe/ui/StartActivity.kt]: Navigation.findNavController() doesnt't work with the BaseActivity. Or I don't know how. It seems to need `this@StartActivity`

File: app/src/main/java/io/gnosis/safe/ui/StartActivity.kt [hunk 6]
```diff
@@ -32,24 +34,33 @@ import pm.gnosis.utils.asEthereumAddress
...
     private val toolbar by lazy { findViewById<View>(R.id.toolbar) }
     private val navBar by lazy { findViewById<BottomNavigationView>(R.id.nav_bar) }
 
+    var comingFromBackground = false
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_start)
 
+        if (settingsHandler.requirePasscodeToOpen && settingsHandler.usePasscode) {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/StartActivity.kt:57]: could you move it to onResume and extend then when statement there? Navigation.findNavController(this@StartActivity,... should be moved to a private method for a better readability (enterPasscode() ?)

Comment (biafra23) [reply app/src/main/java/io/gnosis/safe/ui/StartActivity.kt:57]: Calling it from onResume is too often. We need it once when the app is started  (onCreate) and then when it comes back from the background (activity count was zero)

File: app/src/main/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsFragment.kt [hunk 7]
```diff
@@ -60,7 +63,13 @@ class TransactionDetailsFragment : BaseViewBindingFragment<FragmentTransactionDe
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
-
+        if (requirePasscode && !passcodeShown) {
+            Navigation.findNavController(requireActivity(), R.id.nav_host).navigate(R.id.enterPasscodeFragment, Bundle().apply {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/transactions/details/TransactionDetailsFragment.kt:67]: findNavController() should be enough (insead of Navigation.findNavController(requireActivity(), R.id.nav_host))
Also I would move the whole nav code to a private method for a better readability (setupPasscode() or something like that)

File: app/src/main/java/io/gnosis/safe/ui/settings/app/passcode/PasscodeSettingsFragment.kt [hunk 8]
```diff
@@ -124,8 +124,8 @@ class PasscodeSettingsFragment : SafeOverviewBaseFragment<FragmentSettingsAppPas
         super.onResume()
         with(binding) {
             usePasscode.settingSwitch.isChecked = settingsHandler.usePasscode
-            requireForConfirmations.settingSwitch.isChecked = settingsHandler.requireForConfirmations
-            requireToOpen.settingSwitch.isChecked = settingsHandler.requireToOpen
+            requireForConfirmations.settingSwitch.isChecked = settingsHandler.requirePasscodeForConfirmations
+            requireToOpen.settingSwitch.isChecked = settingsHandler.requirePasscodeToOpen
             useBiometrics.settingSwitch.isChecked = settingsHandler.useBiometrics
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/settings/app/passcode/PasscodeSettingsFragment.kt:129]: could you hide this switch in this pr?

File: app/src/main/java/io/gnosis/safe/ui/StartActivity.kt [hunk 9]
```diff
@@ -58,6 +66,19 @@ class StartActivity : BaseActivity(), SafeOverviewNavigationHandler {
         setupNav()
 
         handleIntent(intent)
+
+        (application as? HeimdallApplication)?.registerForAppState(this)
+
+        if (settingsHandler.askForPasscodeSetupOnFirstLaunch) {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/StartActivity.kt:72]: this is handled in onNewIntent

**Pull request merged**